[Inductor][CPP] Enhance cppcsevar data type deduce

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130317
* #122961
* #130892
* __->__ #130827

**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
[Inductor][CPP] Enhance the cppcsevar data type deduce

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Enhance cppcsevar data type deduce"


**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -535,22 +536,32 @@ def update_on_args(self, name, args, kwargs):
-            self.dtype = torch.bool
+        else:
+            # Induce output dtype by inputs' dtype
+            assert all(
+                arg.dtype is not None for arg in args if isinstance(arg, CppCSEVariable)
+            )
+            self.dtype = functools.reduce(
+                torch.promote_types,  # type: ignore[arg-type]
+                [arg.dtype for arg in args if isinstance(arg, CppCSEVariable)],
+            )
```
Can we move this logic into `_deduce_output_dtype_by_name`?

leslie-fang-intel replied:
Thanks for the comment. As synced offline, move it into a util function inside `cpp_utils.py`

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -535,22 +536,32 @@ def update_on_args(self, name, args, kwargs):
+            )
+        ) is not None:
+            self.dtype = output_dtype
+        elif name == "masked":
+            assert (
+                hasattr(V.interpreter, "current_node")
+                and V.interpreter.current_node.target.startswith("masked_subblock")
+                and get_current_node_opt_ctx() is not None
+            )
             self.dtype = get_current_node_opt_ctx().dtype
```
Perhaps just set the dtype per the current node context regardless of whether it is `masked` or not?

leslie-fang-intel replied:
Perhaps we can also deduce the `masked` data type by inputs' cse variable of other. I have added a TODO here, let's check it later.

jgong5 commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -987,6 +991,52 @@ def sign(x):
 CppOverrides._initialize_pointwise_overrides("cpp")
 
 
+def get_promote_dtype(args):
+    return (
+        functools.reduce(
+            torch.promote_types,  # type: ignore[arg-type]
+            [n.dtype for n in args if isinstance(n, CppCSEVariable)],
+        )
+        if all(n.dtype is not None for n in args if isinstance(n, CppCSEVariable))
+        else None  # not enough info to calculate the promote dtype
+    )
+
+
+def promote_vec_args(new_args):
```
Shall we move them to cpp_utils.py?

leslie-fang-intel replied:
Sure, move it into cpp_utils.py.

leslie-fang-intel committed:
Update on "[Inductor][CPP] Enhance cppcsevar data type deduce"


**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Enhance cppcsevar data type deduce"


**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Enhance cppcsevar data type deduce"


**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/common.py:
```diff
@@ -306,6 +306,47 @@ def boolean_ops():
 }
 
 
+def _deduce_output_dtype_by_name(
```
It is used by other python files. Better to name it public.
```suggestion
def deduce_output_dtype_by_name(
```

leslie-fang-intel replied:
Thanks for the comment and changed.

jgong5 commented on torch/_inductor/codegen/cpp_utils.py:
```diff
@@ -76,6 +86,143 @@
+    )
+
+
+def promote_vec_args(new_args):
```
What's the diff from promoting normal `CppCSEVariable` args that could be either scalars or vectors?

leslie-fang-intel replied:
yes, I think we can do the `promote_args` for `scalars` and `vectors` unified and changed.

leslie-fang-intel committed:
Update on "[Inductor][CPP] Enhance cppcsevar data type deduce"


**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor][CPP] Enhance cppcsevar data type deduce"


**Summary**
Previously, we used `data_type_propagation` at the start of `codegen` to deduce the data type of each node and save this information in `node.meta[OptimizationContext.key]`. Then, we used this node metadata to update the cppcsevar data type in `update_on_args`. However, this method is not always correct. For example, in the codegen of `indirect_indexing` (see [here](https://github.com/pytorch/pytorch/blob/096dc444cee3ebf8c1868ceb92ca4d6ede9f9cc1/torch/_inductor/codegen/common.py#L1844)), we insert nodes on the fly and reuse the node of `indirect_indexing` to set the `cppcsevar` data type. In this PR, we plan to enhance the `cppcsevar` data type deduction:

- We will deduce the `cppcsevar` data type in `update_on_args` by reusing the code in `data_type_propagation`.

- To align the data type of scalar and vector variables, we previously always cast the scalar to the vector's data type. This caused a data type misalignment between `codegen` and `data_type_propagation`. We should use the same data type promotion logic to align the data types of scalar and vector variables.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @lezcano, could you kindly help to take a look of this PR? thanks.

leslie-fang-intel commented:
@jansel could you help to take a look of this PR?

leslie-fang-intel commented:
@pytorchbot merge
